python/vowpalwabbit/sklearn_vw.py

```diff
@@ -398,45 +520,42 @@ def fit(self, X, y=None, sample_weight=None):
             So pipeline can call transform() after fit
 
         """
-        if self.convert_to_vw_:
-            X = tovw(x=X, y=y, sample_weight=sample_weight, convert_labels=self.convert_labels)
 
-        model = self.get_vw()
-
-        # add examples to model
-        for n in range(self.passes_):
-            if n >= 1:
-                X_ = shuffle(X)
-            else:
-                X_ = X
-            for idx, x in enumerate(X_):
-                model.learn(x)
-        self.fit_ = True
-        return self
-
-    def transform(self, X, y=None):
-        """Transform does nothing by default besides closing the model. Transform is required for any estimator
-         in a sklearn pipeline that isn't the final estimator
+        params = {k: v for k, v in self.get_params().items() if v is not None}
 
-        Parameters
-        ----------
-
-        X : {array-like, sparse matrix}, shape (n_samples, n_features or 1 if not convert_to_vw) or
-            Training vector, where n_samples in the number of samples and
-            n_features is the number of features.
-            if not using convert_to_vw, X is expected to be a list of vw formatted feature vector strings with labels
-        y : array-like, shape (n_samples,), optional if not convert_to_vw
-            Target vector relative to X.
+        passes = 1
+        use_data_file = params.get('data', False)
+        use_data_file = params.get('d', use_data_file)
+        if not use_data_file:
+            # remove passes from vw params since we're feeding in the data manually
+            passes = params.pop('passes', passes)
+            if params.get('bfgs', False):
+                raise RuntimeError(
+                    'An external data file must be used to fit models using the bfgs option'
+                )
 
-        Returns
-        -------
+        # remove estimator attributes from vw params
+        params.pop('convert_to_vw', None)
+        params.pop('convert_labels', None)
+        self.vw_ = pyvw.vw(**params)
+
+        if X is not None:
+            if self.convert_to_vw:
+                X = tovw(x=X,
+                         y=y,
+                         sample_weight=sample_weight,
+                         convert_labels=self.convert_labels)
+
+            # add examples to model
+            for n in range(passes):
+                if n >= 1:
+                    examples = shuffle(X)
+                else:
+                    examples = X
+                for idx, example in enumerate(examples):
+                    self.vw_.learn(example)
 
-        X : {array-like, sparse matrix}
-            To be passed into next estimator in pipeline
-        """
-        if not self.get_vw().finished:
-            self.get_vw().finish()
```

Comment:
Are we going to restore this or not? I think It may cause problems.

I am getting this -
```julia
Exception ignored in: <bound method vw.__del__ of <vowpalwabbit.pyvw.vw object at 0x7f29b9b02d50>>
Traceback (most recent call last):
  File "/home/rituraj/.local/lib/python3.6/site-packages/vowpalwabbit-8.8.0-py3.6-linux-x86_64.egg/vowpalwabbit/pyvw.py", line 258, in __del__
    self.finish()
  File "/home/rituraj/.local/lib/python3.6/site-packages/vowpalwabbit-8.8.0-py3.6-linux-x86_64.egg/vowpalwabbit/pyvw.py", line 249, in finish
    if not self.finished:
AttributeError: 'vw' object has no attribute 'finished'
```

Reply:
we should not finish the vw instance until we no longer want to use it (i.e. it's ready for deletion) as any further calls to save will cause a segfault.

I'm not sure why this is used in a transformer if it's not transforming anything? we might want to test out use of the estimator in a pipeline, particularly chained with a downstream estimator to make sure the behavior is as expected.

Until we can determine if this is needed I would exclude it, if we do in fact need the transform method then we need to make sure it works without vw.finish() otherwise check_estimator will crash.